nslsii/areadetector/xspress3.py

```diff
@@ -0,0 +1,913 @@
+        # the original values will be replaced when it is unstaged
+        self.stage_sigs[self.blocking_callbacks] = 1
+        self.stage_sigs[self.enable] = 1
+        self.stage_sigs[self.compression] = "zlib"
+        self.stage_sigs[self.file_template] = "%s%s_%6.6d.h5"
+
+        # allow for a pause when staging this detector
+        self.stage_sleep_time = stage_sleep_time
+
+        self._filestore_res = None
+        # JL: what are mds_keys?
+        self.mds_keys = {
+            channel.channel_number: mds_key_format.format(
+                self=self, channel_number=channel.channel_number
+            )
+            for channel in parent.iterate_channels()
+        }
+
+    def stop(self, success=False):
+        super_stop_return = super().stop(success=success)
+        self.capture.put(0)
+        return super_stop_return
+
+    def kickoff(self):
+        raise NotImplementedError()
+
+    def collect(self):
+        raise NotImplementedError()
+
+    def make_filename(self):
+        fn, rp, write_path = super().make_filename()
+        if self.parent.make_directories.get():
+            makedirs(write_path)
+        return fn, rp, write_path
+
+    def stage(self):
+        logger.debug("staging Xspress3 '%s'", self.parent.prefix)
+
+        # force the Xspress3 to stop acquiring
+        self.parent.cam.acquire.put(0, wait=True)
+
+        total_points_reading = self.parent.total_points.get()
+        if total_points_reading < 1:
+            raise RuntimeError(
+                f"total_points '{self.parent.total_points}' must be set"
+            )
+        spectra_per_point_reading = self.parent.spectra_per_point.get()
+        total_capture = total_points_reading * spectra_per_point_reading
+
+        # stop previous acquisition
+        # JL: this was set above, don't set it again
+        # self.stage_sigs[self.parent.cam.acquire] = 0
+
+        # re-order the stage signals and disable the calc record which is
+        # interfering with the capture count
+        # JL: can we get rid of num_capture_calc_disable?
+        # JL: is self.num_capture even in stage_sigs?
+        self.stage_sigs.pop(self.num_capture, None)
+        self.stage_sigs.pop(self.parent.cam.num_images, None)
+        self.stage_sigs[self.num_capture_calc_disable] = 1
+
+        external_trig_reading = self.parent.external_trig.get()
+        if external_trig_reading:
+            logger.debug(
+                "Xspress3 '%s' will be triggered externally", self.parent.prefix
+            )
+            self.stage_sigs[self.parent.cam.trigger_mode] = "TTL Veto Only"
+            self.stage_sigs[self.parent.cam.num_images] = total_capture
+        else:
+            logger.debug(
+                "Xspress3 '%s' will be triggered internally", self.parent.prefix
+            )
+            self.stage_sigs[self.parent.cam.trigger_mode] = "Internal"
+            # JL: why not total_capture as above?
```

Comment:
The "external" case is for fly scarns where we need to tell the detector the total number of images to be captured in one "acquire" so it needs to know the total number of spectra to be collected, but in the internal case we need to tell it how many images to take per-point for step scans so `trigger` works as expected!